server/channels/app/password/hashers/hashers.go

```diff
@@ -0,0 +1,122 @@
+// This means that the older password hashers can *never* be removed, unless all
+// users whose passwords are not migrated are either forced to re-login, or
+// forced to generate a new password.
+package hashers
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/mattermost/mattermost/server/v8/channels/app/password/parser"
+)
+
+// PasswordHasher is a password hasher compliant with the PHC string format:
+// https://github.com/P-H-C/phc-string-format/blob/master/phc-sf-spec.md
+//
+// Implementations of this interface need to make sure that all of the methods
+// are thread-safe.
+type PasswordHasher interface {
+	// Hash computes a hash of the provided password, returning a PHC string
+	// containing all the information that was needed to compute it: the function
+	// used to generate it, its version, parameters and salt, if needed.
+	//
+	// If an implementation of Hash needs a salt to generate the hash, it will
+	// create one, so that callers don't need to provide it.
+	Hash(password string) (string, error)
+
+	// CompareHashAndPassword compares the parsed PHC and a provided password,
+	// returning an error if and only if the hashes do not match.
+	//
+	// Implementations need to make sure that the comparisons are done in constant
+	// time; e.g., using crypto/internal/fips140/subtle.ConstantTimeCompare.
+	CompareHashAndPassword(hash parser.PHC, password string) error
+}
+
+const (
+	// Maximum password length for all password hashers
+	PasswordMaxLengthBytes = 72
+)
+
+var (
+	// LatestHasher is the hasher currently in use.
+	// Any password hashed with a different hasher must be migrated to this one.
+	LatestHasher PasswordHasher = DefaultPBKDF2()
+
+	// ErrPasswordTooLong is the error returned when the provided password is
+	// longer than [PasswordMaxLengthBytes].
+	ErrPasswordTooLong = fmt.Errorf("password too long; maximum length in bytes: %d", PasswordMaxLengthBytes)
+
+	// ErrMismatchedHashAndPassword is the error returned when the provided
+	// password does not match the stored hash
+	ErrMismatchedHashAndPassword = fmt.Errorf("hash and password do not match")
+)
+
+// getDefaultHasher returns a [BCrypt] hasher, which is somewhat special, since it
+// uses the whole PHC string as the hash.
+func getDefaultHasher(phcString string) (PasswordHasher, parser.PHC) {
+	return NewBCrypt(), parser.PHC{Hash: phcString}
+}
+
+// GetHasherFromPHC returns the password hasher that was used to generate the
+// provided PHC string.
+// If the PHC string is not valid, or the function ID is unknown, this function
+// defaults to the first hasher ever used, which did not properly implement PHC:
+// the [BCrypt] hasher.
+func GetHasherFromPHCString(phcString string) (PasswordHasher, parser.PHC) {
+	phc, err := parser.New(strings.NewReader(phcString)).Parse()
+	if err != nil {
+		// If the PHC string is invalid, return the default hasher
+		return getDefaultHasher(phcString)
+	}
+
+	switch phc.Id {
+	case PBKDF2FunctionId:
+		return LatestHasher, phc
```

Comment:
If we want to modify the parameters for a hash, but keep the function, how does that work?

Reply:
Ah, that's a good question! At first I had added an `IsPHCValid(phc parser.PHC) bool` to the `PasswordHasher` interface, that allowed for hashers to validate that PHC conformed to their function and parameters. I then removed it (not completely, you can still see an unexposed implementation in `PBKDF2`) in the name of simplicity, but that left us without a path to migrate parameters. Will go through this again, thanks!